packages/ui-extensions-sdk/src/utils/security.ts

```diff
@@ -22,76 +19,51 @@ const isRequestKind = (timing: PerformanceResourceTiming) => {
-    previouslyLoadedResources: LoadedResourceIds
-): [LoadedResourceMetaDataBatch, LoadedResourceIds] => {
-    // Use resource timing api to to retrieve application loaded resources
-    const snapshot = performance.getEntriesByType('resource');
-
-    const resources = snapshot
-        .filter(isResourceTimingEntry)
-        // to avoid collecting the same resource twice, we use
-        // the Set difference: snapshot - previousSnapshot
-        .filter(r => !previouslyLoadedResources.has(r.name))
-        .map(r => {
-            // safe because r.name represents the resolved URL of the requested resource.
-            const urlHostname = new URL(r.name).hostname;
-            const ts = performance.timeOrigin;
-            let url;
-            if (!isRequestKind(r)) {
-                // for privacy reason we do not capture full URLs of network calls
-                url = r.name;
-            }
-            return {
-                startTimeTs: ts + r.startTime,
-                secureConnectionStartTs: ts + r.secureConnectionStart,
-                url,
-                urlHostname,
-                initiatorType: r.initiatorType,
-                nextHopProtocol: r.nextHopProtocol,
-                duration: r.duration,
-                decodedBodySize: r.decodedBodySize
-            };
-        });
-
-    return [{ resources }, new Set(snapshot.map(r => r.name))];
+export const parseResource = (
+    r: PerformanceResourceTiming
+): LoadedResourceMetadata => {
+    // safe because r.name represents the resolved URL of the requested resource.
+    const urlHostname = new URL(r.name).hostname;
+    const ts = performance.timeOrigin;
+    let url;
+    if (!isRequestKind(r)) {
+        // for privacy reason we do not capture full URLs of network calls
+        url = r.name;
+    }
+    return {
+        startTimeTs: ts + r.startTime,
+        secureConnectionStartTs: ts + r.secureConnectionStart,
+        url,
+        urlHostname,
+        initiatorType: r.initiatorType,
+        nextHopProtocol: r.nextHopProtocol,
+        duration: r.duration,
+        decodedBodySize: r.decodedBodySize
+    };
 };
 
 export const startResourceMonitoring = (
     reportUsage: (batch: LoadedResourceMetaDataBatch) => void
 ): (() => void) => {
-    let loadedResourceIds: LoadedResourceIds = new Set();
-
     if (!isPerformanceObjectSupported()) {
         return () => {};
     }
 
-    const doResourceCollection = () => {
-        // collect batch of resource-loading data
-        const [batch, newIds] = collectResourceUsage(loadedResourceIds);
-
-        // update index of loaded resources
-        loadedResourceIds = newIds;
-
-        // send to web-ui
-        reportUsage(batch);
+    const observerCallback = (entries: PerformanceObserverEntryList) => {
+        const resources = entries
+            .getEntries()
+            .filter(isResourceTimingEntry)
+            .map(parseResource);
+        return reportUsage({ resources });
```

Comment:
This is really clean, but I think we're going to need a hybrid approach because of the resources that load before the SDK instantiates. So maybe we:

call `performance.getTimingEntries()` once on instantiation, then subscribe the observer for all continuing monitoring?